pilot/pkg/serviceregistry/kube/controller/endpointcontroller.go

```diff
@@ -113,37 +113,40 @@ func updateEDS(c *Controller, epc kubeEndpointsController, ep interface{}, event
 	c.xdsUpdater.EDSUpdate(c.clusterID, string(host), ns, endpoints)
 }
 
-// getPod fetches a pod by IP address.
+// getPod fetches a pod by name or IP address.
 // A pod may be missing (nil) for two reasons:
 // * It is an endpoint without an associated Pod. In this case, expectPod will be false.
 // * It is an endpoint with an associate Pod, but its not found. In this case, expectPod will be true.
 //   this may happen due to eventually consistency issues, out of order events, etc. In this case, the caller
 //   should not precede with the endpoint, or inaccurate information would be sent which may have impacts on
 //   correctness and security.
 func getPod(c *Controller, ip string, ep *metav1.ObjectMeta, targetRef *v1.ObjectReference, host host.Name) (rpod *v1.Pod, expectPod bool) {
-	pod := c.pods.getPodByIP(ip)
-	if pod != nil {
-		return pod, false
-	}
-	// This means, the endpoint event has arrived before pod event.
-	// This might happen because PodCache is eventually consistent.
 	if targetRef != nil && targetRef.Kind == "Pod" {
 		key := kube.KeyFunc(targetRef.Name, targetRef.Namespace)
-		// There is a small chance getInformer may have the pod, but it hasn't
-		// made its way to the PodCache yet as it a shared queue.
-		podFromInformer, f, err := c.pods.informer.GetStore().GetByKey(key)
-		if err != nil || !f {
-			log.Debugf("Endpoint without pod %s %s.%s error: %v", ip, ep.Name, ep.Namespace, err)
+		pod := c.pods.getPodByKey(key)
+		if pod == nil {
+			// This means, the endpoint event has arrived before pod event.
+			// This might happen because PodCache is eventually consistent.
+			log.Debugf("Endpoint without pod %s %s.%s", ip, ep.Name, ep.Namespace)
 			endpointsWithNoPods.Increment()
 			if c.metrics != nil {
 				c.metrics.AddMetric(model.EndpointNoPod, string(host), "", ip)
 			}
 			// Tell pod cache we want to queue the endpoint event when this pod arrives.
 			epkey := kube.KeyFunc(ep.Name, ep.Namespace)
 			c.pods.queueEndpointEventOnPodArrival(epkey, ip)
-			return nil, true
 		}
-		pod = podFromInformer.(*v1.Pod)
+		return pod, true
+	}
+	// This means the endpoint is manually controlled
+	// TODO: this may be not correct because of the hostnetwork pods may have same ip address
+	// Do we have a way to get the pod from only endpoint?
+	pod := c.pods.getPodByIP(ip)
+	if pod != nil {
+		// This prevents selecting a pod in another different namespace
+		if pod.Namespace != ep.Namespace {
+			pod = nil
```

Comment:
When we fail here, we are now missing the `endpointsWithNoPods` logic on L130-L137. This means if the endpoint is later created it will remain broken

Reply:
When it comes in this clause, this is the endpoint created manually not by k8s controllers. In most case, these endpoint has no associated pods.  For the other very rare case(may be < 0.01%), we could not know whether this ep has an associated pod or not.  I donot mind enqueue endpoint on pod arrival in this case, but have a concern about when we should delete the element if there is no pod at all.